File: acp_ui_flutter/lib/core/dag/team_dag.dart
/// Team DAG - DAG 执行引擎
///
/// 支持并行/顺序/混合执行策略

import 'package:flutter/foundation.dart';

/// 执行策略
enum ExecutionStrategy {
  parallel,
  sequential,
  hybrid,
}

/// 节点状态
enum NodeStatus {
  pending,
  waitingSync,
  running,
  completed,
  failed,
  skipped,
}

/// 执行状态
enum ExecutionStatus {
  planning,
  pendingSync,
  running,
  paused,
  completed,
  failed,
  cancelled,
}

/// 团队成员
class TeamMember {
  final String agentId;
  final String templateId;
  final String role;
  final List<String> capabilities;
  final List<String> dependencies;

  TeamMember({
    required this.agentId,
    required this.templateId,
    required this.role,
    this.capabilities = [],
    this.dependencies = [],
  });

  Map<String, dynamic> toJson() => {
    'agent_id': agentId,
    'template_id': templateId,
    'role': role,
    'capabilities': capabilities,
    'dependencies': dependencies,
  };
}

/// 同步点
class SyncPoint {
  final String id;
  final String name;
  final List<String> waitFor;
  final String triggerAction;
  final int timeoutMs;

  SyncPoint({
    required this.id,
    required this.name,
    this.waitFor = [],
    this.triggerAction = '',
    this.timeoutMs = 30000,
  });

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'wait_for': waitFor,
    'trigger_action': triggerAction,
    'timeout_ms': timeoutMs,
  };
}

/// DAG 节点
class DAGNode {
  final String id;
  final String memberId;
  final List<String> dependencies;
  final NodeStatus status;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final String? result;

  DAGNode({
    required this.id,
    required this.memberId,
    this.dependencies = [],
    this.status = NodeStatus.pending,
    this.startedAt,
    this.completedAt,
    this.result,
  });

  factory DAGNode.fromJson(Map<String, dynamic> json) {
    return DAGNode(
      id: json['id'] as String,
      memberId: json['member_id'] as String,
      dependencies: (json['dependencies'] as List?)?.map((e) => e as String).toList() ?? [],
      status: _parseNodeStatus(json['status'] as String),
      startedAt: json['started_at'] != null
          ? DateTime.parse(json['started_at'] as String)
          : null,
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'] as String)
          : null,
      result: json['result'] as String?,
    );
  }

  static NodeStatus _parseNodeStatus(String s) {
    return NodeStatus.values.firstWhere(
      (e) => e.name == s.replaceAll('_', ''),
      orElse: () => NodeStatus.pending,
    );
  }
}

/// 执行计划
class ExecutionPlan {
  final String id;
  final String teamId;
  final ExecutionStrategy strategy;
  final List<DAGNode> nodes;
  final List<SyncPoint> syncPoints;
  final String? currentSyncPoint;
  final ExecutionStatus status;
  final DateTime createdAt;
  final DateTime? startedAt;
  final DateTime? completedAt;

  ExecutionPlan({
    required this.id,
    required this.teamId,
    required this.strategy,
    this.nodes = [],
    this.syncPoints = [],
    this.currentSyncPoint,
    this.status = ExecutionStatus.planning,
    required this.createdAt,
    this.startedAt,
    this.completedAt,
  });

  factory ExecutionPlan.fromJson(Map<String, dynamic> json) {
    return ExecutionPlan(
      id: json['id'] as String,
      teamId: json['team_id'] as String,
      strategy: _parseStrategy(json['strategy'] as String),
      nodes: (json['nodes'] as List?)?.map((e) => DAGNode.fromJson(e as Map<String, dynamic>)).toList() ?? [],
      syncPoints: (json['sync_points'] as List?)?.map((e) => SyncPoint(
        id: e['id'] as String,
        name: e['name'] as String,
        waitFor: (e['wait_for'] as List?)?.map((x) => x as String).toList() ?? [],
        triggerAction: e['trigger_action'] as String? ?? '',
        timeoutMs: e['timeout_ms'] as int? ?? 30000,
      )).toList() ?? [],
      currentSyncPoint: json['current_sync_point'] as String?,
      status: _parseExecutionStatus(json['status'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
      startedAt: json['started_at'] != null
          ? DateTime.parse(json['started_at'] as String)
          : null,
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'] as String)
          : null,
    );
  }

  static ExecutionStrategy _parseStrategy(String s) {
    return ExecutionStrategy.values.firstWhere(
      (e) => e.name == s,
      orElse: () => ExecutionStrategy.hybrid,
    );
  }

  static ExecutionStatus _parseExecutionStatus(String s) {
    return ExecutionStatus.values.firstWhere(
      (e) => e.name == s.replaceAll('_', ''),
      orElse: () => ExecutionStatus.planning,
    );
  }

  double get progress {
    if (nodes.isEmpty) return 0.0;
    final completed = nodes.where((n) => n.status == NodeStatus.completed).length;
    return completed / nodes.length;
  }
}

/// DAG 执行引擎
class DAGEngine {
  final Map<String, ExecutionPlan> _plans = {};

  /// 创建执行计划
  ///
  /// 通过 WebSocket 调用 Tauri 命令 create_dag_plan
  Future<ExecutionPlan> createPlan(
    String teamId,
    List<TeamMember> members,
    List<SyncPoint> syncPoints,
    ExecutionStrategy strategy,
  ) async {
    // TODO: WebSocket 调用
    // final response = await _wsService.sendCommand({
    //   'command': 'create_dag_plan',
    //   'args': {
    //     'team_id': teamId,
    //     'members': members.map((m) => m.toJson()).toList(),
    //     'sync_points': syncPoints.map((s) => s.toJson()).toList(),
    //     'strategy': strategy.name,
    //   },
    // });
    // final plan = ExecutionPlan.fromJson(response);
    // _plans[plan.id] = plan;
    // return plan;

    // 本地创建（备用）
    final planId = 'plan-${DateTime.now().millisecondsSinceEpoch}';
    final nodes = members.map((m) => DAGNode(
      id: 'node-${m.agentId}',
      memberId: m.agentId,
      dependencies: m.dependencies,
      status: NodeStatus.pending,
    )).toList();

    final plan = ExecutionPlan(
      id: planId,
      teamId: teamId,
      strategy: strategy,
      nodes: nodes,
      syncPoints: syncPoints,
      createdAt: DateTime.now(),
    );
    _plans[planId] = plan;
    return plan;
  }

  /// 获取计划进度
  Future<double> getProgress(String planId) async {
    // TODO: WebSocket 调用
    // final response = await _wsService.sendCommand({
    //   'command': 'get_dag_plan_progress',
    //   'args': {'plan_id': planId},
    // });
    // return (response as num).toDouble();

    final plan = _plans[planId];
    return plan?.progress ?? 0.0;
  }

  /// 获取计划
  ExecutionPlan? getPlan(String planId) {
    return _plans[planId];
  }

  /// 获取就绪节点
  List<DAGNode> getReadyNodes(ExecutionPlan plan) {
    final completedIds = plan.nodes
        .where((n) => n.status == NodeStatus.completed)
        .map((n) => n.memberId)
        .toSet();

    return plan.nodes.where((n) {
      if (n.status != NodeStatus.pending) return false;
      return n.dependencies.every((d) => completedIds.contains(d));
    }).toList();
  }

  /// 开始执行节点
  void startNode(String planId, String nodeId) {
    final plan = _plans[planId];
    if (plan == null) return;

    final nodeIdx = plan.nodes.indexWhere((n) => n.id == nodeId);
    if (nodeIdx == -1) return;

    final node = plan.nodes[nodeIdx];
    plan.nodes[nodeIdx] = DAGNode(
      id: node.id,
      memberId: node.memberId,
      dependencies: node.dependencies,
      status: NodeStatus.running,
      startedAt: DateTime.now(),
    );
  }

  /// 完成节点
  void completeNode(String planId, String nodeId, String result) {
    final plan = _plans[planId];
    if (plan == null) return;

    final nodeIdx = plan.nodes.indexWhere((n) => n.id == nodeId);
    if (nodeIdx == -1) return;

    final node = plan.nodes[nodeIdx];
    plan.nodes[nodeIdx] = DAGNode(
      id: node.id,
      memberId: node.memberId,
      dependencies: node.dependencies,
      status: NodeStatus.completed,
      startedAt: node.startedAt,
      completedAt: DateTime.now(),
      result: result,
    );
  }

  /// 检查计划是否完成
  bool isPlanComplete(ExecutionPlan plan) {
    return plan.nodes.every((n) =>
      n.status == NodeStatus.completed || n.status == NodeStatus.skipped
    );
  }

  /// 列出所有计划
  List<ExecutionPlan> listPlans() {
    return _plans.values.toList();
  }
}

File: src-tauri/src/team_dag.rs
//! Team Orchestration - DAG execution engine with SyncPoints
//!
//! Implements parallel/sequential/hybrid execution strategies for multi-agent teams

use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};
use std::collections::{HashMap, HashSet};

/// Execution strategy
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ExecutionStrategy {
    Parallel,    // Execute all agents simultaneously
    Sequential,  // Execute agents one by one
    Hybrid,      // Mix of parallel and sequential based on dependencies
}

impl ExecutionStrategy {
    pub fn as_str(&self) -> &'static str {
        match self {
            ExecutionStrategy::Parallel => "parallel",
            ExecutionStrategy::Sequential => "sequential",
            ExecutionStrategy::Hybrid => "hybrid",
        }
    }
}

/// Team member definition
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamMember {
    pub agent_id: String,
    pub template_id: String,
    pub role: String,           // "frontend", "backend", "qa", "lead"
    pub capabilities: Vec<String>,
    pub dependencies: Vec<String>,  // IDs of agents this one depends on
}

/// Sync point for coordination
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncPoint {
    pub id: String,
    pub name: String,
    pub wait_for: Vec<String>,   // Agent IDs that must complete first
    pub trigger_action: String,  // Action to take when sync point reached
    pub timeout_ms: u64,         // Maximum wait time
}

/// DAG node for execution plan
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DAGNode {
    pub id: String,
    pub member_id: String,
    pub dependencies: Vec<String>,
    pub status: NodeStatus,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub result: Option<String>,
}

/// Node status
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum NodeStatus {
    Pending,
    WaitingSync,
    Running,
    Completed,
    Failed,
    Skipped,
}

impl NodeStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            NodeStatus::Pending => "pending",
            NodeStatus::WaitingSync => "waiting_sync",
            NodeStatus::Running => "running",
            NodeStatus::Completed => "completed",
            NodeStatus::Failed => "failed",
            NodeStatus::Skipped => "skipped",
        }
    }
}

/// Team execution plan
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExecutionPlan {
    pub id: String,
    pub team_id: String,
    pub strategy: ExecutionStrategy,
    pub nodes: Vec<DAGNode>,
    pub sync_points: Vec<SyncPoint>,
    pub current_sync_point: Option<String>,
    pub status: ExecutionStatus,
    pub created_at: DateTime<Utc>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
}

/// Execution status
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ExecutionStatus {
    Planning,
    PendingSync,
    Running,
    Paused,
    Completed,
    Failed,
    Cancelled,
}

impl ExecutionStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            ExecutionStatus::Planning => "planning",
            ExecutionStatus::PendingSync => "pending_sync",
            ExecutionStatus::Running => "running",
            ExecutionStatus::Paused => "paused",
            ExecutionStatus::Completed => "completed",
            ExecutionStatus::Failed => "failed",
            ExecutionStatus::Cancelled => "cancelled",
        }
    }
}

/// DAG execution engine
pub struct DAGEngine {
    plans: HashMap<String, ExecutionPlan>,
    completed_nodes: HashMap<String, HashSet<String>>,
}

impl DAGEngine {
    pub fn new() -> Self {
        Self {
            plans: HashMap::new(),
            completed_nodes: HashMap::new(),
        }
    }

    /// Create execution plan from team definition
    pub fn create_plan(
        &self,
        team_id: &str,
        members: Vec<TeamMember>,
        sync_points: Vec<SyncPoint>,
        strategy: ExecutionStrategy,
    ) -> Result<ExecutionPlan, String> {
        // Validate DAG (no circular dependencies)
        self.validate_dag(&members)?;

        // Build nodes from members
        let nodes: Vec<DAGNode> = members.iter().map(|m| {
            DAGNode {
                id: uuid::Uuid::new_v4().to_string(),
                member_id: m.agent_id.clone(),
                dependencies: m.dependencies.clone(),
                status: NodeStatus::Pending,
                started_at: None,
                completed_at: None,
                result: None,
            }
        }).collect();

        let plan = ExecutionPlan {
            id: uuid::Uuid::new_v4().to_string(),
            team_id: team_id.to_string(),
            strategy,
            nodes,
            sync_points,
            current_sync_point: None,
            status: ExecutionStatus::Planning,
            created_at: Utc::now(),
            started_at: None,
            completed_at: None,
        };

        Ok(plan)
    }

    /// Validate DAG topology (no circular dependencies)
    fn validate_dag(&self, members: &[TeamMember]) -> Result<(), String> {
        // Build adjacency list
        let mut graph: HashMap<String, Vec<String>> = HashMap::new();

        for member in members {
            graph.insert(member.agent_id.clone(), member.dependencies.clone());
        }

        // DFS to detect cycles
        let mut visited = HashSet::new();
        let mut rec_stack = HashSet::new();

        for member in members {
            if self.has_cycle(&member.agent_id, &graph, &mut visited, &mut rec_stack) {
                return Err("Circular dependency detected in team DAG".to_string());
            }
        }

        Ok(())
    }

    /// DFS cycle detection
    fn has_cycle(
        &self,
        node: &str,
        graph: &HashMap<String, Vec<String>>,
        visited: &mut HashSet<String>,
        rec_stack: &mut HashSet<String>,
    ) -> bool {
        if rec_stack.contains(node) {
            return true; // Cycle found
        }

        if visited.contains(node) {
            return false; // Already processed
        }

        visited.insert(node.to_string());
        rec_stack.insert(node.to_string());

        if let Some(deps) = graph.get(node) {
            for dep in deps {
                if self.has_cycle(dep, graph, visited, rec_stack) {
                    return true;
                }
            }
        }

        rec_stack.remove(node);
        false
    }

    /// Get nodes ready to execute (dependencies satisfied)
    pub fn get_ready_nodes<'a>(&self, plan: &'a ExecutionPlan) -> Vec<&'a DAGNode> {
        let completed: HashSet<&'a String> = plan.nodes.iter()
            .filter(|n| n.status == NodeStatus::Completed)
            .map(|n| &n.member_id)
            .collect();

        plan.nodes.iter()
            .filter(|n| {
                n.status == NodeStatus::Pending &&
                n.dependencies.iter().all(|d| completed.contains(d))
            })
            .collect()
    }

    /// Start executing a node
    pub fn start_node(&mut self, plan_id: &str, node_id: &str) -> Result<(), String> {
        let plan = self.plans.get_mut(plan_id)
            .ok_or_else(|| format!("Plan {} not found", plan_id))?;

        let node = plan.nodes.iter_mut()
            .find(|n| n.id == node_id)
            .ok_or_else(|| format!("Node {} not found", node_id))?;

        node.status = NodeStatus::Running;
        node.started_at = Some(Utc::now());

        Ok(())
    }

    /// Complete a node
    pub fn complete_node(&mut self, plan_id: &str, node_id: &str, result: String) -> Result<(), String> {
        let plan = self.plans.get_mut(plan_id)
            .ok_or_else(|| format!("Plan {} not found", plan_id))?;

        let node = plan.nodes.iter_mut()
            .find(|n| n.id == node_id)
            .ok_or_else(|| format!("Node {} not found", node_id))?;

        node.status = NodeStatus::Completed;
        node.completed_at = Some(Utc::now());
        node.result = Some(result);

        // Track completed nodes
        let completed = self.completed_nodes.entry(plan_id.to_string())
            .or_default();
        completed.insert(node.member_id.clone());

        // Check sync points inline
        Self::check_sync_points_inline(plan)?;

        Ok(())
    }

    /// Check sync points (static inline version)
    fn check_sync_points_inline(plan: &mut ExecutionPlan) -> Result<(), String> {
        // Find sync point to trigger
        let sync_point_id = plan.sync_points.iter()
            .find(|sync_point| {
                sync_point.wait_for.iter().all(|agent_id| {
                    plan.nodes.iter()
                        .any(|n| n.member_id == *agent_id && n.status == NodeStatus::Completed)
                })
            })
            .map(|s| s.id.clone());

        if let Some(id) = sync_point_id {
            plan.current_sync_point = Some(id.clone());
            plan.status = ExecutionStatus::PendingSync;

            // Auto-continue
            for node in &mut plan.nodes {
                if node.status == NodeStatus::WaitingSync {
                    node.status = NodeStatus::Pending;
                }
            }

            plan.current_sync_point = None;
            plan.status = ExecutionStatus::Running;
        }

        Ok(())
    }

    /// Check if all nodes are completed
    pub fn is_plan_complete(&self, plan: &ExecutionPlan) -> bool {
        plan.nodes.iter().all(|n|
            n.status == NodeStatus::Completed || n.status == NodeStatus::Skipped
        )
    }

    /// Get execution progress
    pub fn get_progress(&self, plan: &ExecutionPlan) -> f64 {
        let total = plan.nodes.len();
        let completed = plan.nodes.iter()
            .filter(|n| n.status == NodeStatus::Completed)
            .count();

        if total == 0 {
            0.0
        } else {
            completed as f64 / total as f64
        }
    }

    /// Register a plan
    pub fn register_plan(&mut self, plan: ExecutionPlan) {
        self.plans.insert(plan.id.clone(), plan);
    }

    /// Get plan by ID
    pub fn get_plan(&self, plan_id: &str) -> Option<&ExecutionPlan> {
        self.plans.get(plan_id)
    }
}

impl Default for DAGEngine {
    fn default() -> Self {
        Self::new()
    }
}